docs/CANARY_CONFIGURATIONS.md
# LH Canary Configurations

<!-- TOC -->
* [LH Canary Configurations](#lh-canary-configurations)
  * [What is LH Canary?](#what-is-lh-canary)
  * [Metronome](#metronome)
    * [Kafka Topics](#kafka-topics)
    * [Configurations](#configurations)
      * [`lh.canary.metronome.enable`](#lhcanarymetronomeenable)
      * [`lh.canary.metronome.run.frequency.ms`](#lhcanarymetronomerunfrequencyms)
      * [`lh.canary.metronome.run.threads`](#lhcanarymetronomerunthreads)
      * [`lh.canary.metronome.run.requests`](#lhcanarymetronomerunrequests)
      * [`lh.canary.metronome.get.frequency.ms`](#lhcanarymetronomegetfrequencyms)
      * [`lh.canary.metronome.get.threads`](#lhcanarymetronomegetthreads)
      * [`lh.canary.metronome.get.retries`](#lhcanarymetronomegetretries)
      * [`lh.canary.metronome.data.path`](#lhcanarymetronomedatapath)
      * [`lh.canary.metronome.beat.extra.tags.<additional tag>`](#lhcanarymetronomebeatextratagsadditional-tag)
    * [Kafka Configurations](#kafka-configurations)
    * [LH Client Configurations](#lh-client-configurations)
  * [Task Worker](#task-worker)
    * [Kafka Topics](#kafka-topics-1)
    * [Configurations](#configurations-1)
      * [`lh.canary.metronome.worker.enable`](#lhcanarymetronomeworkerenable)
    * [Kafka Configurations](#kafka-configurations-1)
    * [LH Client Configurations](#lh-client-configurations-1)
  * [Aggregator](#aggregator)
    * [Metrics](#metrics)
    * [Kafka Topics](#kafka-topics-2)
    * [Configurations](#configurations-2)
      * [`lh.canary.aggregator.enable`](#lhcanaryaggregatorenable)
      * [`lh.canary.aggregator.store.retention.ms`](#lhcanaryaggregatorstoreretentionms)
      * [`lh.canary.metrics.port`](#lhcanarymetricsport)
      * [`lh.canary.metrics.path`](#lhcanarymetricspath)
      * [`lh.canary.metrics.common.tags.<additional tag>`](#lhcanarymetricscommontagsadditional-tag)
    * [Kafka Configurations](#kafka-configurations-2)
  * [Common Configurations](#common-configurations)
      * [`lh.canary.id`](#lhcanaryid)
      * [`lh.canary.topic.name`](#lhcanarytopicname)
      * [`lh.canary.topic.creation.enable`](#lhcanarytopiccreationenable)
      * [`lh.canary.topic.creation.timeout.ms`](#lhcanarytopiccreationtimeoutms)
      * [`lh.canary.topic.replicas`](#lhcanarytopicreplicas)
      * [`lh.canary.topic.partitions`](#lhcanarytopicpartitions)
      * [`lh.canary.workflow.creation.enable`](#lhcanaryworkflowcreationenable)
      * [`lh.canary.workflow.name`](#lhcanaryworkflowname)
      * [`lh.canary.workflow.version`](#lhcanaryworkflowversion)
      * [`lh.canary.workflow.revison`](#lhcanaryworkflowrevison)
  * [Using Env Variables](#using-env-variables)
<!-- TOC -->

## What is LH Canary?

A Canary is a species of bird that was used in the past by miners to detect toxic fumes.

The LH Canary is a **Synthetic Monitoring Application** that emulates the paths users might take when
using a LH Cluster. It is useful to detect bugs, failures, and outages.

LH Canary take inspiration on **Strimzi Canary**. Strimzi Canary is a part of the Strimzi project that sits outside your
Kafka cluster
and simulates activity to identify problems from a user perspective. It can be used to raise alerts when a Kafka cluster
is not operating as expected. The Canary tool creates a dedicated “Canary” topic with partitions on every broker, sends
messages to the Canary Topic. Each message has a payload containing an ID and a timestamp. The Canary tool also consumes
the messages and measures the latency (via the timestamp), and exposes such metrics in the Prometheus format.

The LH Canary is highly similar to the Strimzi Canary, but it is slightly more complex. Whereas the Strimzi Canary just
needs to produce messages to a topic and make sure they go through, the LH Canary emulates an E2E use case where a
client requests to run a workflow and then a worker executes the scheduled task.

As you can see in the following image, LH Canary has three components an `Aggregator`, a `Metronome` and
a `LH TaskWorker`.

![](../docs/images/lh-canary.png)

## Metronome

It's an application that executes
a [RunWf](https://github.com/littlehorse-enterprises/littlehorse/blob/master/schemas/service.proto#L75)
in a constant interval (ex: every 500ms).
It is configurable to simulate different scenarios.

### Kafka Topics

- `canary-beats` use for producing the metric beats.

### Configurations

#### `lh.canary.metronome.enable`

Flag to enable the metronome component.

- **Type:** boolean
- **Default:** true
- **Importance:** high

---

#### `lh.canary.metronome.run.frequency.ms`

Time between run wf requests.

- **Type:** int
- **Default:** 1000
- **Importance:** medium

#### `lh.canary.metronome.run.threads`

Size of run wf requests thread pool.

- **Type:** int
- **Default:** 1
- **Importance:** medium

---

#### `lh.canary.metronome.run.requests`

Number of wf run to request.

- **Type:** int
- **Default:** 1
- **Importance:** medium

---

#### `lh.canary.metronome.get.frequency.ms`

Time between get run wf requests.

- **Type:** int
- **Default:** 1000
- **Importance:** medium

#### `lh.canary.metronome.get.threads`

Size of get run wf requests thread pool.

- **Type:** int
- **Default:** 1
- **Importance:** medium

---

#### `lh.canary.metronome.get.retries`

Number of get wf run retries in case of failure.

- **Type:** int
- **Default:** 1
- **Importance:** medium

---

#### `lh.canary.metronome.data.path`

Local DB path.

- **Type:** string
- **Default:** /tmp/canaryMetronome
- **Importance:** medium

#### `lh.canary.metronome.beat.extra.tags.<additional tag>`

This config is useful to add more tags to prometheus metrics.
For example: `lh.canary.metronome.beat.extra.tags.my_tag=my-value`.

- **Type:** string
- **Default:** null
- **Importance:** low

Limitations: Please note that it is not recommended to change the tag set once the canary has been started.
For more information check https://github.com/prometheus/client_java/issues/696.

### Kafka Configurations

LH Canary supports all kafka configurations. Use the prefix `lh.canary.kafka` and append the kafka config.
Examples:

- For `security.protocol`, use `lh.canary.kafka.security.protocol`.
- For `bootstrap.servers`, use `lh.canary.kafka.bootstrap.servers`.

### LH Client Configurations

LH Canary supports all LH Client configurations. Use the prefix `lh.canary.` and append the LH config.
Change `_` to `.`.

Examples:

- For `LHC_API_HOST`, use `lh.canary.lhc.api.host`. In case of using env variables: `LH_CANARY_LHC_API_HOST`.
- For `LHC_API_PORT`, use `lh.canary.lhc.api.port`. In case of using env variables: `LH_CANARY_LHC_API_PORT`.

## Task Worker

Received and execute tasks previously schedule by the metronome.
This allows worker to calculate the latency since the workflow run was scheduled
to when it was executed by the worker.

### Kafka Topics

- `canary-beats` use for producing the metric beats.

### Configurations

#### `lh.canary.metronome.worker.enable`

Flag to enable the worker component.

- **Type:** boolean
- **Default:** true
- **Importance:** high

### Kafka Configurations

LH Canary supports all kafka configurations. Use the prefix `lh.canary.` and append the kafka config.
Examples

- For `security.protocol`, use `lh.canary.security.protocol`.
- For `bootstrap.servers`, use `lh.canary.bootstrap.servers`.

### LH Client Configurations

LH Canary supports all LH Client configurations. Use the prefix `lh.canary.` and append the LH config.
Change `_` to `.`.

Examples:

- For `LHC_API_HOST`, use `lh.canary.lhc.api.host`. In case of using env variables: `LH_CANARY_LHC_API_HOST`.
- For `LHC_API_PORT`, use `lh.canary.lhc.api.port`. In case of using env variables: `LH_CANARY_LHC_API_PORT`.

## Aggregator

It's a kafka streams application that consumes the metric beats and
aggregates them into average and max metrics.

It exposes a `/metrics`  endpoint that prometheus scrapes.

### Metrics

| Metric                                      | Description                                                                                                  |
|---------------------------------------------|--------------------------------------------------------------------------------------------------------------|
| `canary_task_run_execution_avg`             | Average time elapsed from when a task was scheduled until it was executed by the worker in milliseconds      |
| `canary_task_run_execution_max`             | Maximum time to execute a task by the worker in milliseconds                                                 |
| `canary_task_run_execution_count`           | Total wf run executed by the worker                                                                          |
| `canary_get_wf_run_request_avg`             | Average time of getting a wf run in milliseconds                                                             |
| `canary_get_wf_run_request_max`             | Max time of getting a wf run in milliseconds                                                                 |
| `canary_get_wf_run_request_count`           | Total get executed wf run by the metronome                                                                   |
| `canary_wf_run_request_avg`                 | Average time of requesting a new wf run in milliseconds                                                      |
| `canary_wf_run_request_max`                 | Max time of requesting a new wf run in milliseconds                                                          |
| `canary_wf_run_request_count`               | Total wf run count by the metronome                                                                          |
| `canary_duplicated_task_run_count`          | Number of detected duplicated task. Useful for data integrity, every task scheduled has to have an unique id |
| `canary_get_wf_run_exhausted_retries_count` | Total wf run id count that failed                                                                            |

### Kafka Topics

- `canary-beats` use for producing the metric beats.
- Kafka streams topics:
  ```
  canary-beats
  ${lh.canary.id}-count-changelog
  ${lh.canary.id}-count-repartition
  ${lh.canary.id}-duplicated-task-run-by-server-changelog
  ${lh.canary.id}-duplicated-task-run-by-server-repartition
  ${lh.canary.id}-duplicated-task-run-changelog
  ${lh.canary.id}-latency-changelog
  ${lh.canary.id}-latency-repartition
  ${lh.canary.id}-metrics-changelog
  ${lh.canary.id}-metrics-repartition
  ```

### Configurations

#### `lh.canary.aggregator.enable`

Flag to enable the aggregator component.

- **Type:** boolean
- **Default:** true
- **Importance:** high

---

#### `lh.canary.aggregator.store.retention.ms`

Total kafka streams store retention in milliseconds.

- **Type:** long
- **Default:** 7200000
- **Importance:** medium

---

#### `lh.canary.metrics.port`

Prometheus scrape endpoint port.

- **Type:** int
- **Default:** 4023
- **Importance:** medium

---

#### `lh.canary.metrics.path`

Prometheus scrape endpoint path.

- **Type:** string
- **Default:** /metrics
- **Importance:** medium

---

#### `lh.canary.metrics.common.tags.<additional tag>`

This config is useful to add default tags to prometheus metrics.
For example: `lh.canary.metrics.common.tags.my_tag=my-value`.

- **Type:** string
- **Default:** null
- **Importance:** low

### Kafka Configurations

LH Canary supports all kafka configurations. Use the prefix `lh.canary.` and append the kafka config.
Examples

- For `state.dir`, use `lh.canary.state.dir`
- For `bootstrap.servers`, use `lh.canary.bootstrap.servers`

## Common Configurations

#### `lh.canary.id`

An identifier for LH Canary. Useful for the internal components like: kafka streams, kafka producers and LH workers.

- **Type:** string
- **Default:** canary-default
- **Importance:** high

---

#### `lh.canary.topic.name`

Metrics beats topic. Use by the aggregator for consuming, and for the metronome and worker for producing metric beats.

- **Type:** string
- **Default:** canary-beats
- **Importance:** high

---

#### `lh.canary.topic.creation.enable`

Flag to enable topics creation.

- **Type:** boolean
- **Default:** false
- **Importance:** high

---

#### `lh.canary.topic.creation.timeout.ms`

Total time to wait for creating the metric beats topic in milliseconds.

- **Type:** int
- **Default:** 5000
- **Importance:** high

---

#### `lh.canary.topic.replicas`

Replicas for metric beats topic.

- **Type:** int
- **Default:** 3
- **Importance:** high

---

#### `lh.canary.topic.partitions`

Partitions for metric beats topic.

- **Type:** int
- **Default:** 12
- **Importance:** high

---

#### `lh.canary.workflow.creation.enable`

Flag to enable canary workflow creation.

- **Type:** boolean
- **Default:** false
- **Importance:** high

---

#### `lh.canary.workflow.name`

Workflow name.

- **Type:** string
- **Default:** canary-workflow
- **Importance:** high

---

#### `lh.canary.workflow.version`

Workflow major version.

- **Type:** int
- **Default:** 0
- **Importance:** high

---

#### `lh.canary.workflow.revison`

Workflow reversion.

- **Type:** int
- **Default:** 0
- **Importance:** high

## Using Env Variables

LH Canary has support for env variables.
Properties defined via env variables will override the value of that property defined in file input and default config.

To construct the environment key variable name for server.properties configs, following steps can be followed:-

* Replace `.` with `_`.
* Examples:
    * For `lh.canary.id`, use `LH_CANARY_ID`.
    * For `lh.canary.kafka.bootstrap.servers`, use `LH_CANARY_KAFKA_BOOTSTRAP_SERVERS`.

By default, de log level for the canary is `INFO`. It is possible to change this
setting the env `LOG_LEVEL`.


docs/CLIENT_CONFIGURATIONS.md
# Workers/Clients Configurations
- [Workers/Clients Configurations](#workersclients-configurations)
  - [Client](#client)
    - [`LHC_API_HOST`](#lhc_api_host)
    - [`LHC_API_PORT`](#lhc_api_port)
    - [`LHC_API_PROTOCOL`](#lhc_api_protocol)
    - [`LHC_CA_CERT`](#lhc_ca_cert)
    - [`LHC_CLIENT_CERT`](#lhc_client_cert)
    - [`LHC_CLIENT_KEY`](#lhc_client_key)
    - [`LHC_OAUTH_CLIENT_ID`](#lhc_oauth_client_id)
    - [`LHC_OAUTH_CLIENT_SECRET`](#lhc_oauth_client_secret)
    - [`LHC_OAUTH_ACCESS_TOKEN_URL`](#lhc_oauth_access_token_url)
    - [`LHC_GRPC_KEEPALIVE_TIME_MS`](#lhc_grpc_keepalive_time_ms)
    - [`LHC_GRPC_KEEPALIVE_TIMEOUT_MS`](#lhc_grpc_keepalive_timeout_ms)
  - [LHCTL](#lhctl)
    - [`LHC_OAUTH_SERVER_URL`](#lhc_oauth_server_url)
  - [Worker](#worker)
    - [`LHW_SERVER_CONNECT_LISTENER`](#lhw_server_connect_listener)
    - [`LHW_NUM_WORKER_THREADS`](#lhw_num_worker_threads)
    - [`LHW_TASK_WORKER_ID`](#lhw_task_worker_id)
    - [`LHW_TASK_WORKER_VERSION`](#lhw_task_worker_version)

## Client

### `LHC_API_HOST`

The bootstrap host for the LH Server.

- **Type:** string
- **Default:** localhost
- **Importance:** high

---

### `LHC_API_PORT`

The bootstrap port for the LH Server.

- **Type:** int
- **Default:** 2023
- **Importance:** high

---

### `LHC_API_PROTOCOL`

The bootstrap protocol for the LH Server. Valid values: `PLAINTEXT` and `TLS`.

- **Type:** string
- **Default:** PLAINTEXT
- **Importance:** high

---

### `LHC_CA_CERT`

Optional location of CA Cert file that issued the server side certificates. For TLS and mTLS connection.

- **Type:** path
- **Default:** null
- **Importance:** medium

---

### `LHC_CLIENT_CERT`

Optional location of Client Cert file for mTLS connection.

- **Type:** path
- **Default:** null
- **Importance:** medium

---

### `LHC_CLIENT_KEY`

Optional location of Client Private Key file for mTLS connection.

- **Type:** path
- **Default:** null
- **Importance:** low

---

### `LHC_OAUTH_CLIENT_ID`

Optional OAuth2 Client Id. Used by the Worker to identify itself at an Authorization Server. Client Credentials Flow.

- **Type:** string
- **Default:** null
- **Importance:** low

---

### `LHC_OAUTH_CLIENT_SECRET`

Optional OAuth2 Client Secret. Used by the Worker to identify itself at an Authorization Server. Client Credentials Flow.

- **Type:** string
- **Default:** null
- **Importance:** low

---

### `LHC_OAUTH_ACCESS_TOKEN_URL`

Optional Access Token URL provided by the OAuth Authorization Server. Used by the Worker to obtain a token using client credentials flow.
It is mandatory if `LHC_OAUTH_CLIENT_ID` and `LHC_OAUTH_CLIENT_SECRET` are provided.

- **Type:** url
- **Default:** null
- **Importance:** low

### `LHC_GRPC_KEEPALIVE_TIME_MS`

Time in milliseconds to configure keepalive pings on the grpc client.

- **Type:** int64
- **Default:** 45000 (45 seconds)
- **Importance:** low

### `LHC_GRPC_KEEPALIVE_TIMEOUT_MS`

Time in milliseconds to configure the timeout for the keepalive pings on the grpc client.

- **Type:** int64
- **Default:** 5000 (5 seconds)
- **Importance:** low


## LHCTL

### `LHC_OAUTH_SERVER_URL`

Optional Authorization Server URL. Used by the client to obtain a token using OAuth 2 authorization code credentials flow. It is used by OIDC to discover the server endpoints.

It is mandatory if `LHC_OAUTH_CLIENT_ID` is provided.

- **Type:** url
- **Default:** null
- **Importance:** low

## Worker

### `LHW_SERVER_CONNECT_LISTENER`

LH Server Listener to connect to.

- **Type:** string
- **Default:** PLAIN
- **Importance:** high

---

### `LHW_NUM_WORKER_THREADS`

The number of worker threads to run. It allows you to improve the task execution performance parallelizing the tasks
assigned to this worker.

- **Type:** int
- **Default:** 8
- **Importance:** medium

---

### `LHW_TASK_WORKER_ID`

Unique identifier for the Task Worker. It is used by the LittleHorse cluster to load balance the worker requests across all servers.
Additionally, it is journalled on every `TaskAttempt` run by the Task Worker, so that you can more easily debug where a request was executed from. It is recommended to set this value for production environments.

- **Type:** string
- **Default:** a random value
- **Importance:** medium

---

### `LHW_TASK_WORKER_VERSION`

Optional version identifier. Intended to be informative. Useful when you're running different version of a worker. Along with the `LHW_TASK_WORKER_ID`, this is journalled on every `TaskAttempt`.

- **Type:** string
- **Default:** ""
- **Importance:** low


sdk-python/littlehorse/config.py
import os
import uuid
from pathlib import Path
from typing import Optional, Union, Any
from grpc import CallCredentials, Channel, ChannelCredentials
import grpc
from jproperties import Properties
from littlehorse.auth import (
    OAuthCredentialsProvider,
    MetadataInterceptor,
    AsyncUnaryUnaryMetadataInterceptor,
    AsyncStreamStreamMetadataInterceptor,
    AsyncStreamUnaryMetadataInterceptor,
    AsyncUnaryStreamMetadataInterceptor,
)
from littlehorse.model import LittleHorseStub
from littlehorse.utils import read_binary
import logging

PREFIXES = ("LHC_", "LHW_")
TLS_PROTOCOL = "TLS"
PLAINTEXT_PROTOCOL = "PLAINTEXT"
ALLOWED_PROTOCOLS = (TLS_PROTOCOL, PLAINTEXT_PROTOCOL)

API_HOST = "LHC_API_HOST"
API_PORT = "LHC_API_PORT"
API_PROTOCOL = "LHC_API_PROTOCOL"
CLIENT_CERT = "LHC_CLIENT_CERT"
TENANT_ID = "LHC_TENANT_ID"
CLIENT_KEY = "LHC_CLIENT_KEY"
CA_CERT = "LHC_CA_CERT"
OAUTH_CLIENT_ID = "LHC_OAUTH_CLIENT_ID"
OAUTH_CLIENT_SECRET = "LHC_OAUTH_CLIENT_SECRET"
OAUTH_TOKEN_ENDPOINT_URL = "LHC_OAUTH_ACCESS_TOKEN_URL"
NUM_WORKER_THREADS = "LHW_NUM_WORKER_THREADS"
TASK_WORKER_ID = "LHW_TASK_WORKER_ID"
TASK_WORKER_VERSION = "LHW_TASK_WORKER_VERSION"
GRPC_KEEPALIVE_TIME_MS = "LHC_GRPC_KEEPALIVE_TIME_MS"
GRPC_KEEPALIVE_TIMEOUT_MS = "LHC_GRPC_KEEPALIVE_TIMEOUT_MS"


class ChannelId:
    def __init__(self, server: str, name: str, is_async: bool) -> None:
        self.server = server
        self.is_async = is_async
        self.name = name

    def __eq__(self, __value: object) -> bool:
        return (
            hasattr(__value, "server")
            and hasattr(__value, "is_async")
            and hasattr(__value, "name")
            and self.server == __value.server
            and self.is_async == __value.is_async
            and self.name == __value.name
        )

    def __hash__(self) -> int:
        return hash(str(self))

    def __str__(self) -> str:
        return str(vars(self))


class LHConfig:
    """Littlehorse Client/Worker configuration.
    A property configured using an environment property
    overrides the value provided using a worker.config file.
    """

    _log = logging.getLogger("LHConfig")

    def __init__(self) -> None:
        self.configs = {
            key.upper(): value
            for key, value in os.environ.items()
            if key.startswith(PREFIXES)
        }
        self._opened_channels: dict[ChannelId, Channel] = {}

    def __str__(self) -> str:
        return "\n".join(
            [
                f"{key}={'******' if 'SECRET' in key or 'PASSWORD' in key else value}"
                for key, value in self.configs.items()
            ]
        )

    def load(self, file_path: Union[str, Path]) -> None:
        """Loads configurations from properties file.

        Args:
            file_path (Union[str, Path]): Path to the properties file.
        """
        properties = Properties()
        properties.load(read_binary(file_path), "utf-8")

        new_configs = {
            key.upper(): value.data
            for key, value in properties.items()
            if key.startswith(PREFIXES)
        }

        self.configs = new_configs | self.configs

    def get(self, key: str) -> Optional[str]:
        """Gets a configuration.

        Args:
            key (str): Configuration key.

        Returns:
            Optional[str]: Configuration value or None if it does not exit.
        """
        return self.configs.get(key)

    def get_or_set_default(self, key: str, default: str) -> str:
        """Gets a configuration, or return a default instead. If the configuration
        does not exit, then the default value will be set.

        Args:
            key (str): Configuration key.
            default (str): Default value in case the configuration
            does not exist.

        Returns:
            str: The configuration's value, or the given default value.
        """
        value = self.configs.get(key)

        if value is None:
            self.configs[key] = default
            return default

        return value

    @property
    def bootstrap_server(self) -> str:
        """Returns the LH Bootstrap server address.

        Returns:
            str: Bootstrap server. Default localhost:2023.
        """
        host = self.get_or_set_default(API_HOST, "localhost")
        port = self.get_or_set_default(API_PORT, "2023")
        return f"{host}:{port}"

    @property
    def ca_cert(self) -> Optional[bytes]:
        """Returns the CA Certificates.

        Returns:
            Optional[bytes]: Root certificates as a byte string,
            or None in case it is not configured.
        """
        cert_path = self.get(CA_CERT)
        return None if cert_path is None else read_binary(cert_path)

    @property
    def client_key(self) -> Optional[bytes]:
        """Returns the client certificate key. For MTLS.

        Returns:
            Optional[bytes]: Certificate key as a byte string.
        """
        client_key_path = self.get(CLIENT_KEY)
        return None if client_key_path is None else read_binary(client_key_path)

    @property
    def client_cert(self) -> Optional[bytes]:
        """Returns the client certificate. For MTLS.

        Returns:
            Optional[bytes]: Certificate as a byte string.
        """
        client_cert_path = self.get(CLIENT_CERT)
        return None if client_cert_path is None else read_binary(client_cert_path)

    @property
    def tenant_id(self) -> Optional[str]:
        return self.get(TENANT_ID)

    def is_secure(self) -> bool:
        """Returns True if a secure connection is configured.

        Returns:
            bool: True if a secure connection is expected.
        """
        protocol = self.get_or_set_default(API_PROTOCOL, PLAINTEXT_PROTOCOL).upper()
        if protocol not in ALLOWED_PROTOCOLS:
            raise ValueError(f"Invalid protocol: {protocol}")
        return protocol == TLS_PROTOCOL

    @property
    def task_worker_id(self) -> str:
        """Returns a Task Worker Id id to identify an instance.

        Returns:
            str: A configured Task Worker ID or a random string otherwise.
        """
        random_id = f"worker-{str(uuid.uuid4()).replace('-', '')}"
        return str(self.get_or_set_default(TASK_WORKER_ID, random_id))

    @property
    def oauth_client_id(self) -> Optional[str]:
        """Returns the configured OAuth2 client id. Used for OIDC authorization.

        Returns:
            str: The OAuth client id.
        """
        return self.get(OAUTH_CLIENT_ID)

    def has_authentication(self) -> bool:
        """Returns True if OAuth is configured.

        Returns:
            bool: True if OAuth is configured.
        """
        return bool(self.get(OAUTH_CLIENT_ID))

    @property
    def oauth_client_secret(self) -> Optional[str]:
        """Returns the configured OAuth2 client secret. Used for OIDC authorization.

        Returns:
            str: The OAuth client secret.
        """
        return self.get(OAUTH_CLIENT_SECRET)

    @property
    def oauth_token_endpoint_url(self) -> Optional[str]:
        """Returns the OAuth2 authorization access token endpoint.
        Used by to obtain a token with the client credentials flow.

        Returns:
            str: The OAuth2 authorization server endpoint.
        """
        return self.get(OAUTH_TOKEN_ENDPOINT_URL)

    @property
    def grpc_keepalive_time_ms(self) -> int:
        """Returns the keepalive ping interval for the internal grpc
        client in milliseconds.

        Returns:
            int: The keepalive interval for grpc.
        """
        return int(self.get_or_set_default(GRPC_KEEPALIVE_TIME_MS, "45000"))

    @property
    def grpc_keepalive_timeout_ms(self) -> int:
        """Returns the keepalive ping timeout for the internal grpc
        client in milliseconds.

        Returns:
            int: The keepalive timeout for grpc.
        """
        return int(self.get_or_set_default(GRPC_KEEPALIVE_TIMEOUT_MS, "5000"))

    @property
    def num_worker_threads(self) -> int:
        """Returns the number of worker threads to run.

        Returns:
            int: The number of worker threads to run. Default 8.
        """
        return int(self.get_or_set_default(NUM_WORKER_THREADS, "8"))

    @property
    def worker_version(self) -> str:
        """Returns the version of this worker.

        Returns:
            str: Task Worker Version. Default empty.
        """
        return self.get_or_set_default(TASK_WORKER_VERSION, "")

    def establish_channel(
        self, server: Optional[str] = None, async_channel: bool = False
    ) -> Channel:
        """Open a RPC channel. Returns a new channel.

        Args:
            server (Optional[str], optional): Target, it will use the
            bootstrap server in case of None. Defaults to None.
            async_channel (bool, optional): Defines if the channel
            will use asyncio. Defaults to False.

        Returns:
            Channel: A closable channel. Use 'with' or channel.close().
        """
        server = server or self.bootstrap_server

        channel_args = [
            ("grpc.keepalive_time_ms", self.grpc_keepalive_time_ms),
            ("grpc.keepalive_timeout_ms", self.grpc_keepalive_timeout_ms),
            ("grpc.keepalive_permit_without_calls", True),
            ("grpc.http2.max_pings_without_data", 0),
        ]

        def create_channel(
            target: Optional[str], options: Any, secure_channel: bool
        ) -> Channel:
            credentials = []
            if not self.is_secure():
                self._log.warning("Establishing insecure channel at %s", server)
            elif self.has_authentication():
                credentials = grpc.composite_channel_credentials(
                    get_ssl_config(),
                    get_oauth_config(),
                )
                self._log.debug("Establishing secure channel with OAuth at %s", server)
            else:
                credentials = get_ssl_config()
                self._log.debug("Establishing secure channel at %s", server)

            # https://github.com/grpc/grpc/issues/31442
            async_interceptors = [
                AsyncUnaryUnaryMetadataInterceptor(self.tenant_id),
                AsyncStreamStreamMetadataInterceptor(self.tenant_id),
                AsyncStreamUnaryMetadataInterceptor(self.tenant_id),
                AsyncUnaryStreamMetadataInterceptor(self.tenant_id),
            ]
            if async_channel and secure_channel:

                return grpc.aio.secure_channel(
                    target,
                    credentials,
                    options,
                    interceptors=async_interceptors,
                )
            elif async_channel:
                return grpc.aio.insecure_channel(
                    target,
                    options,
                    interceptors=async_interceptors,
                )
            elif secure_channel:
                return grpc.intercept_channel(
                    grpc.secure_channel(target, credentials, options=channel_args),
                    MetadataInterceptor(self.tenant_id),
                )
            else:
                return grpc.intercept_channel(
                    grpc.insecure_channel(target, options=channel_args),
                    MetadataInterceptor(self.tenant_id),
                )

        def get_ssl_config() -> ChannelCredentials:
            return grpc.ssl_channel_credentials(
                root_certificates=self.ca_cert,
                private_key=self.client_key,
                certificate_chain=self.client_cert,
            )

        def get_oauth_config() -> CallCredentials:
            return grpc.metadata_call_credentials(
                OAuthCredentialsProvider(
                    client_id=self.oauth_client_id,
                    client_secret=self.oauth_client_secret,
                    token_endpoint_url=self.oauth_token_endpoint_url,
                )
            )

        return create_channel(
            server,
            options=channel_args,
            secure_channel=self.is_secure(),
        )

    def stub(
        self,
        server: Optional[str] = None,
        name: str = "default",
        async_channel: bool = False,
    ) -> LittleHorseStub:
        """Gets a gRPC stub for the LH Public API
        on the configured bootstrap server. It creates a new LittleHorseStub,
        but reuse a grpc.Channel.

        Args:
            server (Optional[str], optional): Target, it will use the
            bootstrap server in case of None. Defaults to None.
            name (str, optional): An optional name. Defaults to "default".
            async_channel (bool, optional): Defines if the channel
            will use asyncio. Defaults to False.

        Returns:
            LittleHorseStub: A gRPC stub.
        """
        channel_id = ChannelId(server or self.bootstrap_server, name, async_channel)
        channel = self._opened_channels.get(channel_id)

        if channel is None:
            channel = self.establish_channel(channel_id.server, channel_id.is_async)

            self._opened_channels[channel_id] = channel
        else:
            self._log.debug("Reusing channel %s", channel_id)

        return LittleHorseStub(channel)
